### PR TITLE
GH-721 Adds distribution public key resource

### PR DESCRIPTION
**Pull request by bb-Ricardo**
This will add the `artifactory_distribution_public_key` resource to manage trusted keys for API endpoint `http://<Node>:<Port>/artifactory/api/security/keys/trusted`.

This PR supersedes #722

### CODE REVIEW

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	var distributionPublicKeySchema = map[string]*schema.Schema{
+		"key_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Returns the key id by which this key is referenced in Artifactory.",
+		},
+		"alias": {
+			Type:         schema.TypeString,
+			Required:     true,
+			Description:  "Will be used as a identifier when uploading/retrieving the public key via REST API.",
```
Comment (chb0github) [inline pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go:47]: Don't hate me 🙈 

'will be used as AN identifier'

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		},
+		"fingerprint": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Returns the computed key fingerprint",
+		},
+		"public_key": {
+			Type:             schema.TypeString,
+			Required:         true,
+			Description:      "The Public key to add as trusted distribution GPG key.",
```
Comment (chb0github) [inline pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go:59]: A trusted....

**Review by chb0github — COMMENTED** [review]
Unless something changed, you don't need to check the resty client for error

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		resp, err := m.(util.ProvderMetadata).Client.R().Delete(fmt.Sprintf("%s/%s", DistributionPublicKeysAPIEndPoint, d.Id()))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		if resp.IsError() {
+			return diag.FromErr(fmt.Errorf("unable to delete key: http request failed: %s", resp.Status()))
+		}
+
+		return nil
```
Comment (alexhung) [inline pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go:135]: Pretty sure you need to reset the resource ID manually. At least we used to need to with other resources. e.g. `d.SetId("")`

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+			{
+				Config: keyBasic,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "alias", fmt.Sprintf("foo-alias%d", id)),
+					resource.TestMatchResourceAttr(fqrn, "public_key", regexp.MustCompile(".*-----BEGIN PGP PUBLIC KEY BLOCK-----.*")),
+					resource.TestCheckResourceAttr(fqrn, "fingerprint", "10:16:2c:c5:1c:db:d0:59:ad:86:d3:66:dc:d1:d9:02:65:03:a8:25"),
+					resource.TestCheckResourceAttr(fqrn, "issued_by", "alan <alann@jfrog.com>"),
+				),
+			},
+		},
```
Comment (alexhung) [inline pkg/artifactory/resource/security/resource_artifactory_distribution_public_key_test.go:94]: Please include an import test. Example: https://github.com/jfrog/terraform-provider-artifactory/blob/master/pkg/artifactory/resource/security/resource_artifactory_group_test.go#L37

**Review by alexhung — CHANGES_REQUESTED** [review]

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/jfrog/terraform-provider-shared/packer"
+	"github.com/jfrog/terraform-provider-shared/predicate"
+	"github.com/jfrog/terraform-provider-shared/util"
+)
+
+const DistributionPublicKeysAPIEndPoint = "artifactory/api/security/keys/trusted"
+
+type distributionPublicKeyPayLoad struct {
+	KeyID       string `hcl:"key_id" json:"kid"`
```
[the remark this answers is not in the document]
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: should be fixed in latest commit

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go (hunk 1 above)
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: should be fixed in latest commit

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go (hunk 2 above)
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: should be fixed in latest commit

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		},
+		"issued_by": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Returns the name and eMail address of issuer.",
+		},
+		"valid_until": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Returns the date/time until this GPG key is valid for.",
```
[the remark this answers is not in the document]
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: should be fixed in latest commit

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,11 @@
+## 7.8.0 (May 4, 2023). Tested on Artifactory 7.55.10
+
+FEATURES:
+
+* resource/artifactory_distribution_public_key: Adds new resource to manage distribution public keys which are used to verify signed release bundles
+  PR:     [#722](https://github.com/jfrog/terraform-provider-artifactory/pull/722)
```
[the remark this answers is not in the document]
Comment (bb-Ricardo) [reply CHANGELOG.md]: should be fixed in latest commit

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go (hunk 3 above)
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go:136]: should be fixed in latest commit

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key_test.go (hunk 4 above)
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key_test.go:101]: added in latest commit

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		result := distributionPublicKeyPayLoad{}
+
+		resp, err := m.(util.ProvderMetadata).Client.R().SetBody(keyPost{
+			d.Get("alias").(string),
+			stripTabs(d.Get("public_key").(string)),
+		}).SetResult(&result).Post(DistributionPublicKeysAPIEndPoint)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		if resp.IsError() {
```
[the remark this answers is not in the document]
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: Changed this statement to
```go
                if err != nil {
			return diag.FromErr(err)
		}
		if !resp.IsSuccess() {
```
as `IsError()` returns false for `>=400` and `IsSuccess()` returns true for `>= 200 and <= 299`. This way we avoid errors in misconfigured redirects (happened in our setup before)

[the remark this answers is not in the document]
Comment (alexhung) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: Weird that `IsError()` is behaving that way for you. Resty's code definitely doesn't do that: https://github.com/go-resty/resty/blob/master/response.go#L144

[the remark this answers is not in the document]
Comment (bb-Ricardo) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: If Resty followed all redirects and still ends with an 3XX status then the request should not be considered successful.

I can change it back to the previous `if resp.IsError()` if you prefer to check it that way. Let me know what you think.

[the remark this answers is not in the document]
Comment (alexhung) [reply pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go]: Resty uses Golang's `http` client and that by default follows redirect. We don't apply any redirect policy to either Golang http client or Resty client.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,11 @@
+## 7.8.0 (May 4, 2023). Tested on Artifactory 7.55.10
```
Comment (danielmkn) [inline CHANGELOG.md:1]: The version should be 7.7.0, since the previous PR was closed. Please fix the conflict.

**Review by danielmkn — CHANGES_REQUESTED** [review]
Please resolve the CHANGELOG.md conflict.

File: pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/jfrog/terraform-provider-shared/packer"
+	"github.com/jfrog/terraform-provider-shared/predicate"
+	"github.com/jfrog/terraform-provider-shared/util"
+)
+
+const DistributionPublicKeysAPIEndPoint = "artifactory/api/security/keys/trusted"
+
+type distributionPublicKeyPayLoad struct {
+	KeyID       string `json:"kid"`
+	Alias       string `hcl:"alias" json:"alias"`
```
Comment (alexhung) [inline pkg/artifactory/resource/security/resource_artifactory_distribution_public_key.go:19]: These `hcl` tags are not needed. Please remove.

**Review by alexhung — CHANGES_REQUESTED** [review]

**Comment by bb-Ricardo** [comment]
Just wondering and curious about the last two changes. It would be a matter of two minutes to change it. Instead it's a longer lasting forth and back here inside this PR. Does this have legal reasons or any other reason?

Thank you.

**Comment by danielmkn** [comment]
If you're asking about the `hcl` tag change, no, there are no legal reasons, the tag is just not needed there.

**Review by alexhung — APPROVED** [review]

**Review by danielmkn — APPROVED** [review]

**Pull request merged**